File: hold-it-ios/Views/Stats/CategoryStatsView.swift
//
//  CategoryStatsView.swift
//  hold-it-ios
//

import SwiftUI

struct CategoryStatsView: View {
    let records: [ResistRecord]

    var categoryStats: [(String, String, Int, Double)] {
        // 按 category+emoji 组合分组，再按名称合并
        var merged: [String: (emoji: String, count: Int)] = [:]
        for record in records {
            let key = record.category
            if merged[key] != nil {
                merged[key]?.count += 1
            } else {
                merged[key] = (emoji: record.categoryEmoji, count: 1)
            }
        }
        let total = records.count
        return merged.map { name, info in
            let percentage = total > 0 ? Double(info.count) / Double(total) : 0
            return (info.emoji, name, info.count, percentage)
        }.sorted { $0.2 > $1.2 }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            HStack {
                Image(systemName: "chart.pie.fill")
                    .foregroundStyle(Color.brand)
                Text("分类统计")
                    .font(.headline)
                Spacer()
            }

            if categoryStats.isEmpty {
                Text("暂无数据")
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
            } else {
                VStack(spacing: 12) {
                    ForEach(categoryStats.indices, id: \.self) { index in
                        let stat = categoryStats[index]
                        CategoryBar(
                            emoji: stat.0,
                            name: stat.1,
                            count: stat.2,
                            percentage: stat.3
                        )
                    }
                }
            }
        }
        .padding(16)
        .background(Color.secondarySystemGroupedBackground)
        .cornerRadius(16)
    }
}

struct CategoryBar: View {
    let emoji: String
    let name: String
    let count: Int
    let percentage: Double

    var body: some View {
        VStack(spacing: 6) {
            HStack {
                Text("\(emoji) \(name)")
                    .font(.subheadline)
                Spacer()
                Text("\(count) 次")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

            GeometryReader { geo in
                RoundedRectangle(cornerRadius: 4)
                    .fill(Color.brand.opacity(0.15))
                    .overlay(
                        HStack {
                            RoundedRectangle(cornerRadius: 4)
                                .fill(Color.brand)
                                .frame(width: geo.size.width * percentage)
                            Spacer()
                        }
                    )
            }
            .frame(height: 8)
        }
    }
}

File: hold_it_flutter/lib/views/stats/category_stats_view.dart
import 'package:flutter/material.dart';
import '../../models/models.dart';
import '../../utils/constants.dart';

class CategoryStatsView extends StatelessWidget {
  final List<ResistRecord> records;

  const CategoryStatsView({super.key, required this.records});

  List<Map<String, dynamic>> get _categoryStats {
    final merged = <String, Map<String, dynamic>>{};
    for (final r in records) {
      final key = r.effectiveCategoryID;
      if (merged.containsKey(key)) {
        merged[key]!['count'] = merged[key]!['count'] + 1;
      } else {
        merged[key] = {
          'emoji': r.categoryEmoji,
          'name': r.category,
          'count': 1,
        };
      }
    }
    final total = records.length;
    return merged.values.map((info) {
      final count = info['count'] as int;
      return {
        ...info,
        'percentage': total > 0 ? count / total : 0.0,
      };
    }).toList()
      ..sort((a, b) => (b['count'] as int).compareTo(a['count'] as int));
  }

  @override
  Widget build(BuildContext context) {
    final stats = _categoryStats;

    return Container(
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(16),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        spacing: 16,
        children: [
          Row(
            spacing: 8,
            children: [
              const Icon(Icons.pie_chart, color: AppColors.brand, size: 18),
              const Text('分类统计', style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold)),
            ],
          ),
          if (stats.isEmpty)
            Text('暂无数据', style: TextStyle(fontSize: 14, color: Colors.grey[600]))
          else
            Column(
              spacing: 12,
              children: stats.map((stat) => _CategoryBar(
                emoji: stat['emoji'] as String,
                name: stat['name'] as String,
                count: stat['count'] as int,
                percentage: stat['percentage'] as double,
              )).toList(),
            ),
        ],
      ),
    );
  }
}

class _CategoryBar extends StatelessWidget {
  final String emoji;
  final String name;
  final int count;
  final double percentage;

  const _CategoryBar({
    required this.emoji,
    required this.name,
    required this.count,
    required this.percentage,
  });

  @override
  Widget build(BuildContext context) {
    return Column(
      spacing: 6,
      children: [
        Row(
          children: [
            Text('$emoji $name', style: const TextStyle(fontSize: 14)),
            const Spacer(),
            Text('$count 次', style: TextStyle(fontSize: 12, color: Colors.grey[600])),
          ],
        ),
        LayoutBuilder(
          builder: (context, constraints) {
            return Container(
              height: 8,
              decoration: BoxDecoration(
                color: AppColors.brand.withOpacity(0.15),
                borderRadius: BorderRadius.circular(4),
              ),
              child: Row(
                children: [
                  Container(
                    width: constraints.maxWidth * percentage,
                    decoration: BoxDecoration(
                      color: AppColors.brand,
                      borderRadius: BorderRadius.circular(4),
                    ),
                  ),
                ],
              ),
            );
          },
        ),
      ],
    );
  }
}
